### PR TITLE
HPCC-24517 Python 3.6 fails if passing in a record with > 255 fields

### PR DESCRIPTION
**Pull request by richardkchapman**
Allegedly ok in Python 7 or later

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@mckellyln @ghalliday

File: plugins/py3embed/py3embed.cpp [hunk 1]
```diff
@@ -48,6 +48,11 @@
   #define Py_TPFLAGS_HAVE_ITER 0
 #endif
 
+#if PY_MINOR_VERSION < 7
+  #define USE_CUSTOM_NAMEDTUPLES
+#endif
+  #define USE_CUSTOM_NAMEDTUPLES
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:54]: is this meant to be here?

File: plugins/py3embed/py3embed.cpp [hunk 2]
```diff
@@ -1197,7 +1312,14 @@ class PythonNamedTupleBuilder : public CInterfaceOf<IFieldProcessor>
     PyObject *getTuple(const RtlTypeInfo *type)
     {
         OwnedPyObject mynamedtupletype = sharedCtx ? sharedCtx->getNamedTupleType(type) : globalState.getNamedTupleType(type);
+        Py_ssize_t len = PyList_Size(args);
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:1315]: does not seem to be used

File: plugins/py3embed/py3embed.cpp [hunk 3]
```diff
@@ -394,10 +498,19 @@ static class Python3xGlobalState
         if (!namedtuple)
         {
             namedtupleTypes.setown(PyDict_New());
+#ifdef USE_CUSTOM_NAMEDTUPLES
+            OwnedPyObject temp_namespace(PyDict_New());
+            PyDict_SetItemString(temp_namespace, "__builtins__",  PyEval_GetBuiltins());  // required for import to work
+            checkPythonError();
+            OwnedPyObject ran = PyRun_String(_py36code, Py_file_input, temp_namespace, temp_namespace);
+            checkPythonError();
+            namedtuple.set(PyDict_GetItemString(temp_namespace, "namedtuple"));
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:507]: setown?

File: plugins/py3embed/py3embed.cpp [hunk 4]
```diff
@@ -381,11 +485,11 @@ static class Python3xGlobalState
         // and if we add our own lock we are liable to deadlock as the code within Py_CompileStringFlags may
         // temporarily release then re-acquire the GIL.
         if (!activityContextTupleType)
-            activityContextTupleType.setown(getNamedTupleType("isLocal,numSlaves,numStrands,slave,strand"));
+            activityContextTupleType.setown(getNamedTupleType("isLocal,numSlaves,numStrands,slave,strand", 5));
         return activityContextTupleType.get();
     }
 
-    PyObject *getNamedTupleType(const char *names)
+    PyObject *getNamedTupleType(const char *names, unsigned numFields)
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:492]: numFields not actually used....

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a few comments

File: plugins/py3embed/py3embed.cpp (hunk 1 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:54]: No, just for me testing with 3.7. I will remove when I squash

File: plugins/py3embed/py3embed.cpp (hunk 4 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:492]: No, that was part of my first failed attempt to fix.I will remove.

File: plugins/py3embed/py3embed.cpp (hunk 3 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:507]: I believe set is correct, based on the documentation of PyDict_GetItemString and other uses of it.

File: plugins/py3embed/py3embed.cpp (hunk 2 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:1315]: No, that was part of my first failed attempt to fix.I will remove.

**Comment by mckellyln** [comment]
My simple testing has shown this fix resolves the issue.

File: plugins/py3embed/py3embed.cpp [hunk 5]
```diff
@@ -48,6 +48,10 @@
   #define Py_TPFLAGS_HAVE_ITER 0
 #endif
 
+#if PY_MINOR_VERSION < 7
```
Comment (mckellyln) [inline plugins/py3embed/py3embed.cpp:51]: Should this be a check on major and minor ?
if PY_MAJOR == 3 && PY_MINOR < 7

**Comment by mckellyln** [comment]
@richardkchapman just one small question.

File: plugins/py3embed/py3embed.cpp (hunk 5 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:51]: This plugin is only for python3, so not really needed

Comment (mckellyln) [reply plugins/py3embed/py3embed.cpp:51]: ok, sure.  I was thinking this was the default python pluging going forward, I forgot it was specifically py3embed.

**Pull request merged**